rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 80 - yubikey PIV integration
```

Comment:
Any chance we could make this generic to various kinds of PIV devices?

Reply:
@codingllama Are you aware of any PIV devices except YubiKey? I'd assume that `go-piv` should work with all piv-compatible devices, but I had a hard time finding any piv-compatible device. Most of them support only FIDO.

Reply:
> @codingllama Are you aware of any PIV devices except YubiKey? 

I could not find one either, except Solokey announced an [upcoming one](https://solokeys.com/blogs/news/update-on-our-new-and-upcoming-security-keys). I think we can expect other hardware keys to follow.

> I'd assume that `go-piv` should work with all piv-compatible devices, but I had a hard time finding any piv-compatible device. Most of them support only FIDO.

I don't think this would be the case, because piv-go does not [expressly support non-yubikey piv cards](https://github.com/go-piv/piv-go#non-yubikey-smartcards) due to a lack of test hardware. I think some parts of the implementation are specific to yubikey and it's `libykpiv.so` PIV-compatible module. Specifically, how `piv-go` retrieves available yubikey cards and performs attestation seems to be specific to `libykpiv.so`. However, while `piv-go` does not [expressly support non-yubikey piv cards](https://github.com/go-piv/piv-go#non-yubikey-smartcards) due to a lack of test hardware, but they expect that most functionality should work and are open to PRs to fix integration with them.

Reply:
Another option we have is to lean on the `pkcs#11` standard, which any PIV-compatible hardware key would also support. We can do most of the primary functionality using `pkcs#11`, and lean on `piv-go` only for yubikey pin management, and PIN/Touch policy logic.

We are already using [crypto11](https://pkg.go.dev/github.com/ThalesIgnite/crypto11) for HSM integration.

Reply:
I'm not sure either. :/

I tried some searches but I could only find nonsense. Maybe @r0mant or someone in the dev channels knows?

Reply:
Found this golden discussion comment in solokeys github - https://github.com/solokeys/solo2/discussions/105#discussioncomment-2418512

Reply:
I'm going to work on making a section of the RFD to touch on these issues. I did a lot of research last week into PIV, pkcs#11, fips-201, etc, I just need to get all my thoughts organized so that we can choose the best path forward.